### PR TITLE
Downtimes CRUD

### PR DESCRIPTION
**Pull request by theamrendrasingh**

### CODE REVIEW

File: handlers/routes.go [hunk 1]
```diff
@@ -192,6 +192,13 @@ func Router() *mux.Router {
 	api.Handle("/api/checkins/{api}", authenticated(checkinDeleteHandler, false)).Methods("DELETE")
 	//r.Handle("/checkin/{api}", http.HandlerFunc(checkinHitHandler))
 
+	// API DOWNTIME Routes
+	api.Handle("/api/downtimes/service/{service_id}", authenticated(apiAllDowntimesForServiceHandler, false)).Methods("GET")
```
Comment (guptaankit015) [inline handlers/routes.go]: service id can be accepted in query params

File: types/downtimes/struct.go [hunk 2]
```diff
@@ -12,4 +12,5 @@ type Downtime struct {
 	Failures  int       `gorm:"column:failures" json:"failures"`
 	Start     time.Time `gorm:"index;column:start" json:"start"`
 	End       time.Time `gorm:"column:end" json:"end"`
+	Manual    bool      `gorm:"default:false;column:manual" json:"manual"`
```
Comment (guptaankit015) [inline types/downtimes/struct.go]: this can be part of service object

File: handlers/downtimes.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	if err := downtime.Update(); err != nil {
+		sendErrorJson(err, w, r)
+		return
+	}
+
+	sendJsonAction(downtime, "update", w, r)
+}
+
+func apiDeleteDowntimeHandler(w http.ResponseWriter, r *http.Request) {
+	downtime, err := findDowntime(r)
```
Comment (guptaankit015) [inline handlers/downtimes.go:135]: we dont need find here

File: types/downtimes/struct.go (hunk 2 above)
Comment (theamrendrasingh) [reply types/downtimes/struct.go]: How?

File: handlers/downtimes.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+		return
+	}
+
+	if downtime.End == zeroTime && req.End != zeroTime {
+		s.ManualDowntime = false
+		s.CurrentDowntime = 0
+		s.Online = true
+		s.LastProcessingTime = zeroTime
+
+		if err := s.Update(); err != nil {
```
Comment (guptaankit015) [inline handlers/downtimes.go]: Need to do update with the specific fields here

File: handlers/downtimes.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+	var zeroTime time.Time
+
+	if downtime.End == zeroTime {
+		s.ManualDowntime = true
+		s.CurrentDowntime = downtime.Id
+		s.Online = false
+		s.LastProcessingTime = zeroTime
+
+		if err := s.Update(); err != nil {
```
Comment (guptaankit015) [inline handlers/downtimes.go]: Need to do update with the specific fields here

File: handlers/downtimes.go [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+	downtime.Type = "manual"
+	downtime.Failures = req.Failures
+	downtime.SubStatus = req.SubStatus
+
+	if err := downtime.Update(); err != nil {
+		sendErrorJson(err, w, r)
+		return
+	}
+
+	if req.End == zeroTime {
```
Comment (guptaankit015) [inline handlers/downtimes.go]: Are we going to support this reopening of closed downtime. We can get rid of this scenario and throw exception if downtime end is not null and request tries to set it to null.

File: types/services/database.go [hunk 7]
```diff
@@ -141,14 +141,16 @@ func (s *Service) UpdateOrder() (err error) {
 
 	d := db.Model(s).Where(" id = ? ", s.Id).Updates(updateFields)
 	if err = d.Error(); d.Error() != nil {
-		log.Errorf("[DB ERROR]Failed toservice order : %s %s %s %s", s.Id, s.Name, updateFields, d.Error())
+		fmt.Errorf("[DB ERROR]Failed toservice order : %s %s %s %s", s.Id, s.Name, updateFields, d.Error())
```
Comment (guptaankit015) [inline types/services/database.go]:
```suggestion
		log.Errorf("[DB ERROR]Failed toservice order : %s %s %s %s", s.Id, s.Name, updateFields, d.Error())
```

File: types/services/database.go [hunk 8]
```diff
@@ -208,6 +210,18 @@ func (s *Service) acquireServiceRun() error {
 	return nil
 }
 
+func (s *Service) UpdateSpecificFields(updateFields map[string]interface{}) error {
+	d := db.Model(s).Where(" id = ?", s.Id).Updates(updateFields)
+	if d.Error() != nil {
+		log.Errorf("[DB ERROR]Failed to update service : %s %s %s %s", s.Id, s.Name, updateFields, d.Error())
```
Comment (guptaankit015) [inline types/services/database.go:216]:
```suggestion
		log.Errorf("[DB ERROR]Failed to update service : %s %s %s %s", s.Id, s.Name, updateFields, d.Error())
		return d.Error()
```

File: types/services/database.go [hunk 9]
```diff
@@ -208,6 +210,18 @@ func (s *Service) acquireServiceRun() error {
...
+	}
+	if d.RowsAffected() == 0 {
+		log.Errorf("[Zero]Failed to update service : %s %s %s %s", s.Id, s.Name, updateFields, d.Error())
```
Comment (guptaankit015) [inline types/services/database.go:220]:
```suggestion
		log.Errorf("[Zero]Failed to update service : %s %s %s %s", s.Id, s.Name, updateFields, d.Error())
		return fmt.Errorf("Failed to update Service fields : %s", s.Id)
```

**Review by guptaankit015 — APPROVED** [review]
Have added few minor comments, please address those before releasing. Rest Changes LGTM.

**Pull request merged**